### DISCUSSION TITLE
[MSSQL] Inserting data in a column with identity fails if the column is not a primary key

### DISCUSSION THREAD

**Discussion by mircea-dogaru-redgate** [sqlalchemy/sqlalchemy#11613] [Usage Questions] [answered]
Recently I came across what seems like a bug. 

We have a table with no primary keys. One of the columns is an integer with identity configured. In our application we need to generate some specific values and insert them into that column.

When we try to insert data into that column using SQL Alchemy, it fails because SQL Alchemy does not enable identity insert. This works fine for other tables with columns with identity if at least one of them is a primary key.

In the codebase it looks like identity is only enabled if one of the columns is a table's autoincrement column. That seems to only be set if a column is a primary key. https://github.com/sqlalchemy/sqlalchemy/blob/e44e805506fa71318e23a2bfad733fbbf5a9ee59/lib/sqlalchemy/dialects/mssql/base.py#L1850

Let me know if I'm missing something.

**Comment by zzzeek** [comment, marked as answer]
this would be something we dont support right now, the "enable identity insert automatically" thing was only to provide some baseline of cross-compatibility between SQL Server's primary key autoincrement scheme and that of other databases like MySQL and SQLite, where sending a specific value for primary key autoincrement is allowed.    However, in the case of SQL Server, if I were writing SQLAlchemy today I wouldn't even have this behavior built in since it's not appropriate as an "automatic" thing, as this is essentially an ALTER TABLE statement that we would assume is database-wide (cc @gordthompson ) and isn't really appropriate for an automatic on-the-fly thing.

So that said if you are inserting special fixture data into an IDENTITY column it's best to put that operation into a block where on the connection you set IDENTITY INSERT off / on for the operation.

**Reply by mircea-dogaru-redgate** [reply]
Thank you for the detailed explanation @zzzeek . Your suggestion is the way we went with as well.

**Reply by MarkJacksonRG** [reply]
Thanks for the answer. However I believe IDENTITY_INSERT is session specific, not database wide. I.e. it's not like ALTER TABLE. The docs do [say this](https://learn.microsoft.com/en-us/sql/t-sql/statements/set-identity-insert-transact-sql?view=sql-server-ver16#remarks), although it's just a single mention of _session_ and it's easy to overlook. So I think it would be fine to use IDENTITY_INSERT in an automatic on the fly thing.

**Reply by zzzeek** [reply]
it's still monumentally inefficient for SET IDENTITY_INSERT ON/OFF to run for every single INSERT statement individually.    I still would not have set this implicitly if I were writing SQLAlchemy today.

**Reply by MarkJacksonRG** [reply]
I see, thanks.

**Reply by CaselIT** [reply]
> it's still monumentally inefficient for SET IDENTITY_INSERT ON/OFF to run for every single INSERT statement individually. I still would not have set this implicitly if I were writing SQLAlchemy today.

would you be open to deprecating it? Or at least make it an engine parameter that people can opt into

**Reply by zzzeek** [reply]
im not really sure as this would really mess up existing applications.   certainly easy to detect when it's happening though

**Reply by CaselIT** [reply]
Then maybe add a new bool parameter in the dialect, with default to warn?
This way existing applications would see the warning but can explicitly opt into this behaviour?

**Reply by zzzeek** [reply]
these are all good ideas I just dont have a good sense of what is the real world use of this behavior so i feel uncertain about it

**Reply by CaselIT** [reply]
let's try that for 2.1 and see how it goes

**Reply by zzzeek** [reply]
alrighty

**Reply by CaselIT** [reply]
tracked at #11620.

**Reply by mircea-dogaru-redgate** [reply]
@zzzeek my team and I are building a tool to allow users to generate data using a variety of data generation techniques for various testing purposes, manual or automated. Because of that we need to be able to insert specific values even in identity columns.

**Reply by CaselIT** [reply]
note that not all db allow you do to that. postgresql requires `OVERRIDING SYSTEM VALUE` on a generated always by identity, that I'm not sure if sqlalchemy supports. Also generated column (not by identity) cannot be written to, even when specifying overriding system value

**Reply by mircea-dogaru-redgate** [reply]
Those are very good points @CaselIT . We're currently focussing on MSSQL. Computed columns are ok to skip, as they're usually generated based on other values. But we'll tackle those use cases later on.
